pcp/PCPDynamicMeter.c

```diff
@@ -0,0 +1,368 @@
+   PCPDynamicMetric *metric;
+   char* p;
+
+   if ((p = strchr(key, '.')) == NULL)
+      return;
+   *p++ = '\0'; /* end the name, p is now the attribute, e.g. 'label' */
+
+   if (String_eq(p, "metric")) {
+      /* lookup a dynamic metric with this name, else create */
+      metric = PCPDynamicMeter_lookupMetric(meters, meter, key);
+
+      /* use derived metrics in dynamic meters for simplicity */
+      char* error;
+      if (pmRegisterDerivedMetric(metric->name, value, &error) < 0) {
+         char note[1024];
+         xSnprintf(note, sizeof(note),
+                   "failed to parse expression in %s at line %u\n%s\n%s",
+                   path, line, error, pmGetProgname());
+         free(error);
+         errno = EINVAL;
+         CRT_fatalError(note);
+      }
+   } else {
+      /* this is a property of a dynamic metric - the metric expression */
+      /* may not have been observed yet - i.e. we allow for any ordering */
+      metric = PCPDynamicMeter_lookupMetric(meters, meter, key);
+      if (String_eq(p, "color")) {
+         if (String_eq(value, "gray"))
+             metric->color = DYNAMIC_GRAY;
+         else if (String_eq(value, "darkgray"))
+             metric->color = DYNAMIC_DARKGRAY;
+         else if (String_eq(value, "red"))
+             metric->color = DYNAMIC_RED;
+         else if (String_eq(value, "green"))
+             metric->color = DYNAMIC_GREEN;
+         else if (String_eq(value, "blue"))
+             metric->color = DYNAMIC_BLUE;
+         else if (String_eq(value, "cyan"))
+             metric->color = DYNAMIC_CYAN;
+         else if (String_eq(value, "magenta"))
+             metric->color = DYNAMIC_MAGENTA;
+         else if (String_eq(value, "yellow"))
+             metric->color = DYNAMIC_YELLOW;
+         else if (String_eq(value, "white"))
+             metric->color = DYNAMIC_WHITE;
+      } else if (String_eq(p, "label")) {
+         free_and_xStrdup(&metric->label, value);
+      } else if (String_eq(p, "suffix")) {
+         free_and_xStrdup(&metric->suffix, value);
+      }
+   }
+}
+
+// Ensure a valid name for use in a PCP metric name and in htoprc
+static void PCPDynamicMeter_validateMeterName(char* key, const char *path, unsigned int line) {
+   char* p = key;
+   char* end = strrchr(key, ']');
+
+   if (end) {
+      *end = '\0';
+   } else {
+      char note[1024];
+      xSnprintf(note, sizeof(note),
+                "No closing brace on meter name at %s line %u\n\"%s\"",
+                path, line, key);
+      errno = EINVAL;
+      CRT_fatalError(note);
+   }
+
+   while (*p) {
+      if (p == key) {
+         if (!isalpha(*p) && *p != '_')
+            break;
+      } else {
+         if (!isalnum(*p) && *p != '_')
+            break;
+      }
+      p++;
+   }
+   if (*p != '\0') { /* badness */
+      char note[1024];
+      xSnprintf(note, sizeof(note),
+                "Invalid meter name at %s line %u\n\"%s\"",
+                path, line, key);
```

Comment:
May crash htop on overlong meter names (from config file?)

Reply:
It'd have to be either the file path that was super long I guess, or the libpcp-supplied error message.  We're about to call CRT_fatalError here, so its mainly a matter of how we fail.  Looking into the xSnprintf internals, we're not going to return from that if we would overrun the buffer - its going to call _exit internally - so there's probably not anything meaningful we can do here.

Reply:
We could use `xAsprintf` here and avoid the whole buffer length issue. Just need to make sure the returned string is freed subsequently.

Reply:
OK, will make that change.  There is no 'subsequently' though - the string is passed to CRT_fatalError.